pkg/pillar/cmd/zedmanager/handleresolve.go

```diff
@@ -0,0 +1,148 @@
+	return &config
+}
+
+func lookupResolveStatus(ctx *zedmanagerContext,
+	key string) *types.ResolveStatus {
+
+	log.Infof("lookupResolveStatus(%s)\n", key)
+	sub := ctx.subAppImgResolveStatus
+	c, _ := sub.Get(key)
+	if c == nil {
+		log.Infof("lookupResolveStatus(%s) not found\n", key)
+		return nil
+	}
+	status := c.(types.ResolveStatus)
+	log.Infof("lookupResolveStatus(%s) Done\n", key)
+	return &status
+}
+
+func handleResolveStatusDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	log.Infof("handleResolveStatusDelete for %s\n", key)
+	ctx := ctxArg.(*zedmanagerContext)
+	status := statusArg.(types.ResolveStatus)
+	config := lookupResolveConfig(ctx, status.Key())
+	if config != nil {
+		log.Infof("handleResolveStatusDelete delete config for %s\n",
+			key)
+		unpublishResolveConfig(ctx, config)
+	}
+	log.Infof("handleResolveStatusDelete done for %s\n", key)
+}
+
+func handleResolveStatusModify(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	log.Infof("handleResolveStatusModify for %s\n", key)
+	ctx := ctxArg.(*zedmanagerContext)
+	rs := statusArg.(types.ResolveStatus)
+	pub := ctx.pubAppInstanceStatus
+	items := pub.GetAll()
+	for _, st := range items {
+		status := st.(types.AppInstanceStatus)
+		if status.RefCount > 0 {
+			log.Infof("Updating images SHA for app instance %v\n",
+				status.UUIDandVersion.UUID)
+			if len(rs.Error) != 0 {
+				errStr := fmt.Sprintf("Error occurred while resolving tags, updating app instance %s.\n",
+					status.Key())
+				log.Errorf(errStr)
+				status.SetError(rs.Error, rs.ErrorSource, rs.ErrorTime)
+				publishAppInstanceStatus(ctx, &status)
+				continue
+			}
+			if rs.ImageSha256 == "" {
+				errStr := fmt.Sprintf("Empty SHA in resolve status, updating app instance %s.\n",
+					status.Key())
+				log.Errorf(errStr)
+				status.SetError(errStr, agentName, time.Now())
+				publishAppInstanceStatus(ctx, &status)
+				continue
+			}
+			config := lookupAppInstanceConfig(ctx, status.Key())
+			if config == nil {
+				errStr := fmt.Sprintf("App Instance config not found for %s while resolving tags\n",
+					status.Key())
+				log.Errorf(errStr)
+				status.SetError(errStr, agentName, time.Now())
+				publishAppInstanceStatus(ctx, &status)
+				continue
+			}
+			for i, sc := range config.StorageConfigList {
+				ss := &status.StorageStatusList[i]
+				if ss.Name != rs.Name && ss.DatastoreID != rs.DatastoreID {
+					continue
+				}
+				if ss.ImageSha256 == "" {
+					ss.ImageSha256 = rs.ImageSha256
+				} else if ss.ImageSha256 != rs.ImageSha256 {
+					MaybeRemoveStorageStatus(ctx, config.UUIDandVersion.UUID, ss)
+					deleteAppAndImageHash(ctx, status.UUIDandVersion.UUID,
+						ss.ImageID)
+					ss.UpdateFromStorageConfig(sc)
+					ss.ImageSha256 = rs.ImageSha256
+				}
+				addAppAndImageHash(ctx, config.UUIDandVersion.UUID,
+					ss.ImageID, ss.ImageSha256)
+				maybeLatchImageSha(ctx, *config, ss)
+			}
+			status.RefCount--
```

Comment:
Clearing this (and the suggested HasResolverRef replacement) should go together unpublishing the ResolverRequest.

Reply:
Updated.